### PR TITLE
1895 - made input keys translateable

### LINKED ISSUE #1895: Add translation support for key names
Key names like "Left", "Right", "Print" can't be translated currently, support for these should be added.

This is in SpecifiedInputKey.cs

### PR DESCRIPTION
**Pull request by FuJa0815**
closes #1895 

closes https://github.com/Revolutionary-Games/Thrive/pull/2062 (as this superceeds that PR).

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
Looks pretty good, assuming you didn't miss any of the key names. Just need to run the localize script and provide the English names for the keys.

**Comment by FuJa0815** [comment]
There are way to many strings right now, this is just an export of all KeyList values.

**Comment by hhyyrylainen** [comment]
> There are way too many strings right now

But shouldn't ALL possible keys have translations?

**Comment by FuJa0815** [comment]
I mean, do strings like "F1" really need a translation?

**Comment by FuJa0815** [comment]
If a key does not exist in the KeyNames.cs file, the `Translate` method just returns the string as it is

**Comment by hhyyrylainen** [comment]
Arguably, those don't, that are just plain like that. But even stuff like the KP (for keypad) prefixed keys should be able to be translated.

**Comment by hhyyrylainen** [comment]
Also for the duplicate code check, if it still complains once you remove the unneeded translations, I think there should be a special comment syntax to ignore the check for that file.

**Comment by FuJa0815** [comment]
Escape is impossible to assign, because Escape cancels the key assignment
And I think Backtab should be recognized as "Shift+Tab" by godot, but I haven't tested this

**Comment by hhyyrylainen** [comment]
I guess backtab is fine then, but if we want to translate the keyprompt for canceling a keybinding, I think having the ESCAPE key name would be nice.

Actually I just got one idea: I think it would make much clearer to translators if the key names were "KEY_HOME" etc. instead of just plain "HOME" as I think if we ever have the same words as the keyboard used elsewhere, we would have the flexibility of different translations for them. I think all that would need is prepending "KEY_" to the uppercased key name and a bit of regex replace on the dummy translations code.

**Comment by FuJa0815** [comment]
Do you mean I should provide the `Translate` function `"KEY_"+code.ToString()` and remove the potential "Key_" from the result of the `Translate` method?

**Comment by hhyyrylainen** [comment]
Oh yea, that makes it a bit more inconvenient, then. I guess we'll just live with having "HOME" meaning the key and if a separate place needs that it will need to be something like "HOME_PLACE"
So yeah, it's fine now. Now just need to tell the duplicate code finder to ignore the code it's complaining about.

**Comment by FuJa0815** [comment]
I'll try, I've actually never coded ruby before, but it doesn't seem that hard :P

**Comment by hhyyrylainen** [comment]
Looks like https://www.jetbrains.com/help/resharper/dupFinder.html#configuring-dupfinder-options has an option called `exclude-by-comment` which should allow configuring a comment that can be placed at the start of a file to exclude it from duplicate finding.

File: locale/en.po [hunk 1]
```diff
@@ -899,15 +899,797 @@ msgstr ""
...
+msgid "NUMLOCK"
+msgstr "Num Lock"
+
+#: ../src/engine/input/key_mapping/KeyNames.cs:141
+msgid "SCROLLLOCK"
+msgstr "Scroll Lock"
+
+#: ../src/engine/input/key_mapping/KeyNames.cs:142
+msgid "SUPERL"
+msgstr "Win Left"
```
Comment (hhyyrylainen) [inline locale/en.po:1425]: This should be named "Super Left", I don't particularly want Windows-only terminology used here.

File: locale/en.po [hunk 2]
```diff
@@ -899,15 +899,797 @@ msgstr ""
...
+
+#: ../src/engine/input/key_mapping/KeyNames.cs:143
+msgid "SUPERR"
+msgstr "Win Right"
```
Comment (hhyyrylainen) [inline locale/en.po:1429]: Same here

File: locale/en.po [hunk 3]
```diff
@@ -899,15 +899,797 @@ msgstr ""
...
+msgid "PAGEUP"
+msgstr "Page Up"
+
+#: ../src/engine/input/key_mapping/KeyNames.cs:138
+msgid "PAGEDOWN"
+msgstr "Page Down"
+
+#: ../src/engine/input/key_mapping/KeyNames.cs:139
+msgid "CAPSLOCK"
+msgstr "Capslock"
```
Comment (hhyyrylainen) [inline locale/en.po:1413]: `Caps Lock` as Wikipedia uses that and it will be more consistent with num and scroll lock.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
This looks good overall, but I'd like a couple of the English translations to be changed.

File: locale/en.po (hunk 1 above)
Comment (FuJa0815) [reply locale/en.po:1425]: I've thought that the "Windows Key" is more widely known, but you're right, Super is the correct term.

**Comment by hhyyrylainen** [comment]
There seems to be an error in the translation files: E 0:00:01.610   load_translation: Expected '"' at end of message while parsing: res://locale/en.po:1040
  <C++ virhe>   Method failed. Returning: RES()
  <C++ lähdekoodi>core/io/translation_loader_po.cpp:133 @ load_translation()

**Comment by hhyyrylainen** [comment]
Might need to just replace the \\ with `Backslash` as it seems that godot's .po parser might be incompatible with the escaped backslash.

**Comment by FuJa0815** [comment]
the conflict popup isn't working yet

File: src/engine/input/key_mapping/InputActionItem.cs [hunk 4]
```diff
@@ -136,7 +136,7 @@ public override int GetHashCode()
         var inputActionItem = (InputActionItem)target.InputActionItemScene.Instance();
 
         inputActionItem.InputName = data.InputName;
-        inputActionItem.DisplayName = data.Name;
+        inputActionItem.DisplayName = TranslationServer.Translate(data.Name);
```
Comment (hhyyrylainen) [inline src/engine/input/key_mapping/InputActionItem.cs:139]: What did this fix?

Comment (hhyyrylainen) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: As far as I can tell, the "data.Name" here is already translated like "Move left"

Comment (FuJa0815) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: somehow, it isn't... I've checked the code, there is a `TranslateFrom` Attribute and the source string exists in the translation file.

Comment (hhyyrylainen) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: What text are you seeing and where that isn't translated. When debugging with a breakpoint on this line, it looks like when initially building the options menu, this is not translated, but afterwards it is. Anyway the real fix is to make sure that NamedInputAction gets translated when the game starts.

Comment (FuJa0815) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: the debugger shows me that Name is always untranslated.

Comment (hhyyrylainen) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: Bug seems to be that SimulationParameters.ApplyTranslations doesn't get called when the game loads. It should be called the first time it is loaded. So maybe in SimulationParameters constructor?

Comment (hhyyrylainen) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: Actually, I remember that a recent PR by @Kasterisk removed the SimulationParameters.ApplyTranslations call when the parameters are loaded. I suspect that, that should be put back in, at least when the options are initially loaded.

Comment (FuJa0815) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: yeah, `ApplyTranslations` only gets called when changing the language and not when the game starts

Comment (hhyyrylainen) [reply src/engine/input/key_mapping/InputActionItem.cs:139]: I just opened a PR trying to address this problem: https://github.com/Revolutionary-Games/Thrive/pull/2175

**Review by hhyyrylainen — APPROVED** [review]
If this now works along with my PR https://github.com/Revolutionary-Games/Thrive/pull/2175 I think this would be nice to get merged now.

**Comment by hhyyrylainen** [comment]
Seems to be fine, testing this again now with that my PR merged. I'll merge this as soon as the CI checks finish, which might take a while as there might be a job queue forming (I plan on addressing this build queue problem somewhat soon).

**Pull request merged**